gvisor: Depends on Go 1.14

###### Checklist

<!-- Note that these are hard requirements -->

<!--
You can use the "Go to file" functionality on github to find the package
Then you can go to the history for this package
Find the latest "package_name: old_version -> new_version" commit
The "new_version" is the the current version of the package
-->
- [x] Checked the [nixpkgs master branch](https://github.com/NixOS/nixpkgs)
<!--
Type the name of your package and try to find an open pull request for the package
If you find an open pull request, you can review it!
There's a high chance that you'll have the new version right away while helping the community!
-->
- [x] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

###### Project name
`nix search` name: nixos.gvisor (gvisor)
<!--
The current version can be found easily with the same process than above for checking the master branch
If an open PR is present for the package, take this version as the current one and link to the PR
-->
current version: 1.14
desired version: 1.16

###### Notify maintainers
<!--
Search your package here: https://search.nixos.org/packages?channel=unstable
If no maintainer is listed for your package, tag the person that last updated the package
-->

maintainers: @andrew-d

###### Note for maintainers

Please tag this issue in your PR.

## Additional

```
$ nix-shell --run "colmena build -f nix/hive.nix"
[INFO ] Enumerating nodes...
[INFO ] Selected all 5 nodes.
     (...) ❌ 0s Evalation failed: Nix exited with error code: 1
[ERROR] Evaluation of 5 nodes failed. Logs:
[ERROR] error: Package ‘go-1.14.15’ in /nix/store/q5cl71pgl7ia32llcr79qmwh4dhpgg8g-nixpkgs-src/pkgs/development/compilers/go/1.14.nix:264 is marked as insecure, refusing to evaluate.
[ERROR] 
[ERROR] 
[ERROR] Known issues:
[ERROR]  - Support for Go 1.14 ended with the release of Go 1.16: https://golang.org/doc/devel/release.html#policy
[ERROR] 
[ERROR] You can install it anyway by allowing this package, using the
[ERROR] following methods:
[ERROR] 
[ERROR] a) To temporarily allow all insecure packages, you can use an environment
[ERROR]    variable for a single invocation of the nix tools:
[ERROR] 
[ERROR]      $ export NIXPKGS_ALLOW_INSECURE=1
[ERROR] 
[ERROR] b) for `nixos-rebuild` you can add ‘go-1.14.15’ to
[ERROR]    `nixpkgs.config.permittedInsecurePackages` in the configuration.nix,
[ERROR]    like so:
[ERROR] 
[ERROR]      {
[ERROR]        nixpkgs.config.permittedInsecurePackages = [
[ERROR]          "go-1.14.15"
[ERROR]        ];
[ERROR]      }
[ERROR] 
[ERROR] c) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
[ERROR]    ‘go-1.14.15’ to `permittedInsecurePackages` in
[ERROR]    ~/.config/nixpkgs/config.nix, like so:
[ERROR] 
[ERROR]      {
[ERROR]        permittedInsecurePackages = [
[ERROR]          "go-1.14.15"
[ERROR]        ];
[ERROR]      }
[ERROR] 
[ERROR] 
[ERROR] (use '--show-trace' to show detailed location information)
[ERROR] 
```

So, this issue is a little more nuanced than I initially thought.

Go 1.14 just doesn't build for me at all.

```
--- FAIL: TestChown (0.00 s)
  os_unix_test.go:58: gid: 100
  os_unix_test.go:70: groups: [ 65534 65534 65534 65534 65534 65534 65534 65534 65534 65534 65534 100 ]
  os_unix_test.go:73: chown /tmp/_Go_TestChown368184601 - 1 65534: chown /tmp/_Go_TestChown368184601: invalid argument
--- FAIL: TestFileChown (0.00s)
  os_unix_test.go:101: gid: 100
  os_unix_test.go:113: groups:  [65534 65534 65534 65534 65534 65534 65534 65534 65534 65534 65534 100]
  os_unix_test.go:116: fchown /tmp/_Go_TestFileChown977451428 -1 65534: chown /tmp/_Go_TestFileChown977451428: invalid argument
--- FAIL: TestLchown (0.00s)
  os_unix_test.go:153: gid: 100
  os_unix_test.go:168: groups:  [65534 65534 65534 65534 65534 65534 65534 65534 65534 65534 65534 100]
  os_unix_test.go:171: lchown /tmp/_Go_TestLchown8193635072 -1 65534: lchown /tmp/_Go_TestLchown8193635072: invalid argument
FAIL
FAIL    os      1.112s
```

Go 1.15 and 1.16 also don't build but for different reasons.

```
ERROR: /build/source/pkg/tcpip/link/rawfile/BUILD:5:11: GoCompilePkg pkg/tcpip/link/rawfile/linux_amd64_pure_stripped/rawfile%/gvisor.dev/gvisor/pkg/tcpip/link/rawfile.a failed (Exit 1): builder failed: error executing command bazel-out/host/bin/external/go_sdk/builder compilepkg -sdk external/go_sdk -installsuffix linux_amd64 -src pkg/tcpip/link/rawfile/blockingpoll_noyield_unsafe.go -src ... (remaining 33 argument(s) skipped)
```

```
compilepkg: error running subcommand: exit status 2
/build/output/sandbox/linux-sandbox/619/execroot/__main__/pkg/tcpip/link/rawfile/rawfile_unsafe.go:137:10: undefined: BlockingPoll
/build/output/sandbox/linux-sandbox/619/execroot/__main__/pkg/tcpip/link/rawfile/rawfile_unsafe.go:159:10: undefined: BlockingPoll
/build/output/sandbox/linux-sandbox/619/execroot/__main__/pkg/tcpip/link/rawfile/rawfile_unsafe.go:189:14: undefined: BlockingPoll
```

--- a/nixos/doc/manual/from_md/release-notes/rl-2111.section.xml
+++ b/nixos/doc/manual/from_md/release-notes/rl-2111.section.xml
@@ -56,6 +56,13 @@
           this version for the entire lifecycle of the 21.11 release.
         </para>
       </listitem>
+      <listitem>
+        <para>
+          Those making use of <literal>buildBazelPackage</literal> will
+          need to regenerate the fetch hashes (preferred), or set
+          <literal>fetchConfigured = false;</literal>.
+        </para>
+      </listitem>
     </itemizedlist>
   </section>
   <section xml:id="other-notable-changes">

--- a/nixos/doc/manual/release-notes/rl-2111.section.md
+++ b/nixos/doc/manual/release-notes/rl-2111.section.md
@@ -24,4 +24,7 @@ In addition to numerous new and upgraded packages, this release has the followin
 * PHP 7.3 is no longer supported due to upstream not supporting this
   version for the entire lifecycle of the 21.11 release.
 
+* Those making use of `buildBazelPackage` will need to regenerate the fetch
+  hashes (preferred), or set `fetchConfigured = false;`.
+
 ## Other Notable Changes

--- a/pkgs/applications/virtualization/gvisor/default.nix
+++ b/pkgs/applications/virtualization/gvisor/default.nix
@@ -1,6 +1,8 @@
 { lib
 , buildBazelPackage
 , fetchFromGitHub
+, callPackage
+, bash
 , cacert
 , git
 , glibcLocales
@@ -9,6 +11,7 @@
 , iptables
 , makeWrapper
 , procps
+, protobuf
 , python3
 }:
 
@@ -16,9 +19,12 @@ let
   preBuild = ''
     patchShebangs .
 
+    substituteInPlace tools/defs.bzl \
+      --replace "#!/bin/bash" "#!${bash}/bin/bash"
+
     # Tell rules_go to use the Go binary found in the PATH
     sed -E -i \
-      -e 's|go_version\s*=\s*"[^"]+",|go_version = "host",|g' \
+      -e 's|go_version\s*=\s*"[^"]+"|go_version = "host"|g' \
       WORKSPACE
 
     # The gazelle Go tooling needs CA certs
@@ -31,20 +37,37 @@ let
     export GOPATH=
   '';
 
+  # Patch the protoc alias so that it always builds from source.
+  rulesProto = fetchFromGitHub {
+    owner = "bazelbuild";
+    repo = "rules_proto";
+    rev = "f7a30f6f80006b591fa7c437fe5a951eb10bcbcf";
+    sha256 = "10bcw0ir0skk7h33lmqm38n9w4nfs24mwajnngkbs6jb5wsvkqv8";
+    extraPostFetch = ''
+      sed -i 's|name = "protoc"|name = "_protoc_original"|' $out/proto/private/BUILD.release
+      cat <<EOF >>$out/proto/private/BUILD.release
+      alias(name = "protoc", actual = "@com_github_protocolbuffers_protobuf//:protoc", visibility = ["//visibility:public"])
+      EOF
+    '';
+  };
+
 in buildBazelPackage rec {
   name = "gvisor-${version}";
-  version = "2019-11-14";
+  version = "20210518.0";
 
   src = fetchFromGitHub {
     owner = "google";
     repo  = "gvisor";
-    rev   = "release-20191114.0";
-    sha256 = "0kyixjjlws9iz2r2srgpdd4rrq94vpxkmh2rmmzxd9mcqy2i9bg1";
+    rev   = "release-${version}";
+    sha256 = "15a6mlclnyfc9mx3bjksnnf4vla0xh0rv9kxdp34la4gw3c4hksn";
   };
 
   nativeBuildInputs = [ git glibcLocales go makeWrapper python3 ];
 
   bazelTarget = "//runsc:runsc";
+  bazelFlags = [
+    "--override_repository=rules_proto=${rulesProto}"
+  ];
 
   # gvisor uses the Starlark implementation of rules_cc, not the built-in one,
   # so we shouldn't delete it from our dependencies.
@@ -76,14 +99,14 @@ in buildBazelPackage rec {
       rm -f "$bazelOut"/java.log "$bazelOut"/java.log.*
     '';
 
-    sha256 = "0fhmlq0d2317gwhma2mz1anb69j4chybk90j71j88wpgw1hxbk34";
+    sha256 = "13pahppm431m198v5bffrzq5iw8m79riplbfqp0afh384ln669hb";
   };
 
   buildAttrs = {
     inherit preBuild;
 
     installPhase = ''
-      install -Dm755 bazel-bin/runsc/*_pure_stripped/runsc $out/bin/runsc
+      install -Dm755 bazel-out/*/bin/runsc/runsc_/runsc $out/bin/runsc
 
       # Needed for the 'runsc do' subcomand
       wrapProgram $out/bin/runsc \

--- a/pkgs/build-support/build-bazel-package/default.nix
+++ b/pkgs/build-support/build-bazel-package/default.nix
@@ -35,7 +35,7 @@ args@{
 # required for the build as configured, rather than fetching all the dependencies
 # which may not work in some situations (e.g. Java code which ends up relying on
 # Debian-specific /usr/share/java paths, but doesn't in the configured build).
-, fetchConfigured ? false
+, fetchConfigured ? true
 
 # Don’t add Bazel --copt and --linkopt from NIX_CFLAGS_COMPILE /
 # NIX_LDFLAGS. This is necessary when using a custom toolchain which

--- a/pkgs/build-support/build-bazel-package/default.nix
+++ b/pkgs/build-support/build-bazel-package/default.nix
@@ -126,7 +126,7 @@ in stdenv.mkDerivation (fBuildAttrs // {
       find $bazelOut/external -maxdepth 1 -type l | while read symlink; do
         name="$(basename "$symlink")"
         rm "$symlink"
-        test -f "$bazelOut/external/@$name.marker" && rm "$bazelOut/external/@$name.marker"
+        test -f "$bazelOut/external/@$name.marker" && rm "$bazelOut/external/@$name.marker" || true
       done
 
       # Patching symlinks to remove build directory reference

--- a/pkgs/development/python-modules/tensorflow/default.nix
+++ b/pkgs/development/python-modules/tensorflow/default.nix
@@ -288,9 +288,9 @@ let
     fetchAttrs = {
       # cudaSupport causes fetch of ncclArchive, resulting in different hashes
       sha256 = if cudaSupport then
-        "1i7z2a7bc2q1vn1h9nx1xc6g1r1cby2xvbcs20fj9h6c2fgaw9j4"
+        "10m6qj3kchgxfgb6qh59vc51knm9r9pkng8bf90h00dnggvv8234"
       else
-        "0s8q5rxq8abr50c5jpwv96ncfc0k8jw7w70ri8viqy031g9v9v45";
+        "04a98yrp09nd0p17k0jbzkgjppxs0yma7m5zkfrwgvr4g0w71v68";
     };
 
     buildAttrs = {

--- a/pkgs/development/tools/bazel-watcher/default.nix
+++ b/pkgs/development/tools/bazel-watcher/default.nix
@@ -56,7 +56,7 @@ buildBazelPackage rec {
       sed -e '/^FILE:@bazel_gazelle_go_repository_tools.*/d' -i $bazelOut/external/\@*.marker
     '';
 
-    sha256 = "0rwwjjj6zaj4hdcbsbp0di53xn6203r2vgpddhdrp8iph9ab60cg";
+    sha256 = "1j175z3d4fbi4pl35py7yjq7ywrvwin6id131jv32hx0ck4g1m46";
   };
 
   buildAttrs = {

--- a/pkgs/servers/http/envoy/default.nix
+++ b/pkgs/servers/http/envoy/default.nix
@@ -3,7 +3,9 @@
 , fetchFromGitHub
 , stdenv
 , cmake
+, gn
 , go
+, jdk
 , ninja
 , python3
 , nixosTests
@@ -15,8 +17,8 @@ let
     # However, the version string is more useful for end-users.
     # These are contained in a attrset of their own to make it obvious that
     # people should update both.
-    version = "1.16.2";
-    commit = "e98e41a8e168af7acae8079fc0cd68155f699aa3";
+    version = "1.17.3";
+    commit = "46bf743b97d0d3f01ff437b2f10cc0bd9cdfe6e4";
   };
 in
 buildBazelPackage rec {
@@ -26,7 +28,7 @@ buildBazelPackage rec {
     owner = "envoyproxy";
     repo = "envoy";
     rev = srcVer.commit;
-    hash = "sha256-aWVMRKFCZzf9/96NRPCP4jiW38DJhXyi0gEqW7uIpnQ=";
+    hash = "sha256:09zzr4h3zjsb2rkxrvlazpx0jy33yn9j65ilxiqbvv0ckaralqfc";
 
     extraPostFetch = ''
       chmod -R +w $out
@@ -36,29 +38,29 @@ buildBazelPackage rec {
     '';
   };
 
-  patches = [
-    # Quiche needs to be updated to compile under newer GCC.
-    # This is a manual backport of https://github.com/envoyproxy/envoy/pull/13949.
-    ./0001-quiche-update-QUICHE-tar-13949.patch
-
-    # upb needs to be updated to compile under newer GCC.
-    # This is a manual backport of https://github.com/protocolbuffers/upb/commit/9bd23dab4240b015321a53c45b3c9e4847fbf020.
-    ./0002-Add-upb-patch-to-make-it-compile-under-GCC10.patch
-  ];
   postPatch = ''
     sed -i 's,#!/usr/bin/env python3,#!${python3}/bin/python,' bazel/foreign_cc/luajit.patch
+    sed -i '/javabase=/d' .bazelrc
+    # Patch paths to build tools, and disable gold because it just segfaults.
+    substituteInPlace bazel/external/wee8.genrule_cmd \
+      --replace '"''$$gn"' '"''$$(command -v gn)"' \
+      --replace '"''$$ninja"' '"''$$(command -v ninja)"' \
+      --replace '"''$$WEE8_BUILD_ARGS"' '"''$$WEE8_BUILD_ARGS use_gold=false"'
   '';
 
   nativeBuildInputs = [
     cmake
     python3
+    gn
     go
+    jdk
     ninja
   ];
 
   fetchAttrs = {
-    sha256 = "0q72c2zrl5vc8afkhkwyalb2h0mxn3133d4b9z4gag0p95wbwgc0";
+    sha256 = "sha256:1cy2b73x8jzczq9z9c1kl7zrg5iasvsakb50zxn4mswpmajkbj5h";
     dontUseCmakeConfigure = true;
+    dontUseGnConfigure = true;
     preInstall = ''
       # Strip out the path to the build location (by deleting the comment line).
       find $bazelOut/external -name requirements.bzl | while read requirements; do
@@ -78,10 +80,14 @@ buildBazelPackage rec {
       >$bazelOut/external/config_validation_pip3/PyYAML-5.3.1-cp38-cp38-linux_x86_64.whl
       >$bazelOut/external/protodoc_pip3/PyYAML-5.3.1-cp38-cp38-linux_x86_64.whl
       >$bazelOut/external/thrift_pip3/thrift-0.13.0-cp38-cp38-linux_x86_64.whl
+
+      # Remove Unix timestamps from go cache.
+      rm -rf $bazelOut/external/bazel_gazelle_go_repository_cache/{gocache,pkg/mod/cache,pkg/sumdb}
     '';
   };
   buildAttrs = {
     dontUseCmakeConfigure = true;
+    dontUseGnConfigure = true;
     dontUseNinjaInstall = true;
     preConfigure = ''
       sed -i 's,#!/usr/bin/env bash,#!${stdenv.shell},' $bazelOut/external/rules_foreign_cc/tools/build_defs/framework.bzl
@@ -98,7 +104,6 @@ buildBazelPackage rec {
     '';
   };
 
-  fetchConfigured = true;
   removeRulesCC = false;
   removeLocalConfigCc = true;
   removeLocal = false;

--- a/pkgs/servers/http/pomerium/default.nix
+++ b/pkgs/servers/http/pomerium/default.nix
@@ -11,15 +11,15 @@ let
 in
 buildGoModule rec {
   pname = "pomerium";
-  version = "0.13.3";
+  version = "0.14.4";
   src = fetchFromGitHub {
     owner = "pomerium";
     repo = "pomerium";
     rev = "v${version}";
-    hash = "sha256-g0w1aIHvf2rJANvGWHeUxdnyCDsvy/PQ9Kp8nDdT/0w=";
+    hash = "sha256:097csr8f43cn0iq030ajvvpwnwcfmjxyyk6pcisdy937axlrzska";
   };
 
-  vendorSha256 = "sha256-grihU85OcGyf9/KKrv87xZonX5r+Z1oHQTf84Ya61fg=";
+  vendorSha256 = "sha256:0n45xvwjiqyh41dsm4z1rnkgkycf5wfmacm804hqnd5rz7xk2shf";
   subPackages = [
     "cmd/pomerium"
     "cmd/pomerium-cli"

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -24160,7 +24160,7 @@ in
   gv = callPackage ../applications/misc/gv { };
 
   gvisor = callPackage ../applications/virtualization/gvisor {
-    go = go_1_14;
+    go = go_1_16;
   };
 
   gvisor-containerd-shim = callPackage ../applications/virtualization/gvisor/containerd-shim.nix { };

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -19098,7 +19098,10 @@ in
 
   engelsystem = callPackage ../servers/web-apps/engelsystem { php = php74; };
 
-  envoy = callPackage ../servers/http/envoy { };
+  envoy = callPackage ../servers/http/envoy {
+    go = go_1_15;
+    jdk = openjdk11;
+  };
 
   etcd = callPackage ../servers/etcd { };
   etcd_3_4 = callPackage ../servers/etcd/3.4.nix { };

